src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexEmbeddedBraceMatcher.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.Common;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.LanguageServices;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.RegularExpressions.LanguageServices
+{
+    /// <summary>
+    /// Brace matching impl for embedded regex strings.
+    /// </summary>
+    internal sealed class RegexEmbeddedBraceMatcher : IEmbeddedBraceMatcher
+    {
+        private readonly RegexEmbeddedLanguage _language;
+
+        public RegexEmbeddedBraceMatcher(RegexEmbeddedLanguage language)
+        {
+            _language = language;
+        }
+
+        public async Task<EmbeddedBraceMatchingResult?> FindBracesAsync(
+            Document document, int position, CancellationToken cancellationToken)
+        {
+            var option = document.Project.Solution.Workspace.Options.GetOption(
+                RegularExpressionsOptions.HighlightRelatedRegexComponentsUnderCursor, document.Project.Language);
+            if (!option)
+            {
+                return null;
+            }
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(position);
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            if (RegexPatternDetector.IsDefinitelyNotPattern(token, syntaxFacts))
+            {
+                return null;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var detector = RegexPatternDetector.TryGetOrCreate(semanticModel, _language);
+            var tree = detector?.TryParseRegexPattern(token, cancellationToken);
+
+            if (tree == null)
+            {
+                return null;
+            }
+
+            return GetMatchingBraces(tree, position);
+        }
+
+        private static EmbeddedBraceMatchingResult? GetMatchingBraces(RegexTree tree, int position)
+        {
+            var virtualChar = tree.Text.FirstOrNullable(vc => vc.Span.Contains(position));
+            if (virtualChar == null)
+            {
+                return null;
+            }
+
+            var ch = virtualChar.Value;
+            if (ch != '(' && ch != ')')
```

Comment:
I don't think `{}` is particular relevant in a regex.  it only happens in something like `\p{Lower}`.  And, in that case, it's not hard to understand where the matching brace is.

`[` is interesting, but didn't feel complex enough to need the matching.  in other words, you don't have much nested or much complexity there.  It's not hard to 'scan' to find the end.  That's very contrasted with parens which are the really compelx way of nesting. 

Let me know what you think.

Reply:
The `TestRegex1` example and its nested brackets made me want it. Perhaps it's not a realistic example though? Are you worried it might get too highlighty?

Reply:
It' a reasonable request.  I'll do it tonight :)  It just wasn't something i felt was critical :)

Reply:
Added support for comments and character classes.